runtime/lua/vim/diagnostic.lua

```diff
@@ -895,22 +901,31 @@ M.handlers.signs = {
     end
 
     local ns = M.get_namespace(namespace)
-    if not ns.user_data.sign_group then
-      ns.user_data.sign_group = string.format('vim.diagnostic.%s', ns.name)
+    if not ns.user_data.sign_ns then
+      local sign_ns = string.format('vim.diagnostic.%s', ns.name)
+      ns.user_data.sign_ns = vim.api.nvim_create_namespace(sign_ns)
     end
 
-    local sign_group = ns.user_data.sign_group
+    local sign_ns = ns.user_data.sign_ns
     for _, diagnostic in ipairs(diagnostics) do
-      vim.fn.sign_place(0, sign_group, sign_highlight_map[diagnostic.severity], bufnr, {
+      local lnum = diagnostic.lnum
+      local sign_text = diagnostic.sign_text or opts.sign_text and opts.sign_text[diagnostic.severity]
```

Comment:
The diagnostic object has no `sign_text` field. If you want to allow users to attach `sign_text` as an arbitrary value it should be in the `user_data` table (which is designed for this purpose).

Also be sure to index into `opts.signs` for signs-specific options.

```suggestion
      local sign_text = diagnostic.user_data.sign_text or (opts.signs and opts.signs.sign_text and opts.signs.sign_text[diagnostic.severity])
```

You may want to split this up into a proper `if` statement...